src/index/index.c

```diff
@@ -0,0 +1,166 @@
+}
+
+void freeKey(void *key) {
+  SIValue_Free(key);
+}
+
+// Create and populate index for specified property
+// (This function will create separate string and numeric indices if property has mixed types)
+void indexProperty(RedisModuleCtx *ctx, const char *graphName, AST_IndexNode *indexOp) {
+  const char *index_label = indexOp->target.label;
+  const char *index_prop = indexOp->target.property;
+  LabelStore *store = LabelStore_Get(ctx, STORE_NODE, graphName, index_label);
+
+  LabelStoreIterator it;
+  LabelStore_Scan(store, &it);
+
+  char *nodeId;
+  uint16_t nodeIdLen;
+  Node *node;
+  EntityProperty *prop;
+
+  Index *index = createIndex(index_label, index_prop);
+
+  Vector_Push(tmp_index_store, index);
+
+  int prop_index = 0;
+  while(LabelStoreIterator_Next(&it, &nodeId, &nodeIdLen, (void**)&node)) {
+    // If the sought property is at a different offset than it occupied in the previous node,
+    // then seek and update
+    if (strcmp(index_prop, node->properties[prop_index].name)) {
+      for (int i = 0; i < node->prop_count; i ++) {
+        prop = node->properties + i;
+        if (!strcmp(index_prop, prop->name)) {
+          prop_index = i;
+          break;
+        }
+      }
+    }
+    prop = node->properties + prop_index;
+    // This value will be cloned within the skiplistInsert routine if necessary
+    SIValue *key = &prop->value;
+
+    if (key->type == T_STRING) {
+      skiplistInsert(index->string_sl, key, node);
+    } else if (key->type & SI_NUMERIC) {
+      skiplistInsert(index->numeric_sl, key, node);
+    } else { // This property was neither a string nor numeric value; raise a run-time error.
+      assert(0);
+    }
+  }
+}
+
+Index* createIndex(const char *label, const char *property) {
+  Index *index = malloc(sizeof(Index));
+  index->target.label = strdup(label);
+  index->target.property = strdup(property);
+
+  index->iter = NULL;
+
+  index->string_sl = skiplistCreate(compareStrings, NULL, compareNodes, cloneKey, freeKey);
+  index->numeric_sl = skiplistCreate(compareNumerics, NULL, compareNodes, cloneKey, freeKey);
+
+  return index;
+}
+
+Index* retrieveIndex(const char *label, const char *property) {
+  // TODO Once indices are stored in the Redis keyspace, we can replace this loop
+  // with a more efficient lookup
+  Index *current_index = NULL;
+  for (int i = 0; i < Vector_Size(tmp_index_store); i ++) {
+    Vector_Get(tmp_index_store, i, &current_index);
+    if (!strcmp(label, current_index->target.label) && !strcmp(property, current_index->target.property)) {
+      return current_index;
+    }
+  }
+  return NULL;
+}
+
+/*
+ * Before the ExecutionPlan has been constructed, we can analyze the FilterTree
+ * to see if a scan operation can employ an index. This function will return a
+ * struct containing an appropriate index (if any) and the boundaries for setting its iterator.
+ */
+IndexBounds selectIndexFromFilters(Vector *filters, const char *label) {
```

Comment:
Once the optimisation step within the execution plan is implemented, this function can be simplified
as we'll be dealing with a single label & property.